init.lua

```diff
@@ -146,19 +175,30 @@ function cudnn.toDescriptor(t)
    return descriptor
 end
 
+function cudnn.createDescriptors(count, descs_type, create_func, destroy_func)
+   local ds = ffi.new(descs_type, count)
+   for i = 0, count - 1 do
+      errcheck(create_func, ds + i)
+   end
+   local function destroyDescriptors(ds)
+      for i = 0, count - 1 do
+         errcheck(destroy_func, ds[i])
+      end
+   end
+   ffi.gc(ds, destroyDescriptors)
+   return ds
+end
 
 local sharedBuffer = {}
-for i=1,numDevices do
-    sharedBuffer[i] = {}
-end
 
 function cudnn.getSharedWorkspace()
     local device = cutorch.getDevice()
-    local stream = cutorch.getStream() -- starts from 0
-    if not sharedBuffer[device][stream] then
-       sharedBuffer[device][stream] = torch.CudaTensor(1)
+    if not sharedBuffer[device] then
```

Comment:
@ngimel: but, this is almost exactly what this implementation is doing:
1) There is an explicit limit added in init.lua, user can set it to any number:
cudnn.maxWorkspaceGPUMemPercent = 95
2) there is no special explicit utility function, but the way state machine works is exactly that: after first iteration (counted from cudnn.reset()), second warmup iteration uses up to the maximum workspace specified above to calculate the biggest findex (it's per device but it's trivial to revert to per-stream option). After this is done, shared workspace for each device is being reset to the maximum calculated.